triplea-game/triplea
# Deployable, tested HTTP Server & Finish Error Upload Feature

## Overview
- Fix up gradle config for http-server to create a launchable artifact.
- Add an integration test for drop wizard http server, fires up the http server and verifies http client code can connect.
- Add database support to http server (per drop wizard recommendation, using JDBI framework)
- Add rate limiting for error report upload. Except for http server deployment, this completes the error report upload feature. For server side rate limiting we add a simple history table and update it on the server side when error reports are submitted. Of note, whenever we add to the history table we'll also purge any records older than two days to prevent the table from growing. An initial rate limit of 5 error reports per day has been configured.
- Update headed-client to have a 'fail-early' check if upload limit has been reached. If so we'll show a dialog explaining this to the user instead of showing the upload-error-report form. To enable this a new GET endpoint was added: '/can-submit-error-report'
- Set a default value for 'github API token' to be 'test'. This is used as a magic value to trigger a stubbed response when creating github issues. To safeguard against using a stub in production, a 'prod' flag is added which will crash the application if we are configured to use a stub response.

## Functional Changes
- Error report upload now has a server side rate limit, 5 per day
- Client side error report upload will tell the user when they hit the limit if they click "report to triplea" when seeing errors

## Manual Testing Performed
- fired up a local http server and tested out the endpoints with curl. (Note, the new integration tests really helped quite a lot, like most automated testing, that testing was more valuable and replaced any curl testing, but it's not bad to have curl testing nonetheless)
- fired up a local http server, lobby, and added a "generate error button" to the homepage that threw an exception. I clicked this to verify the DB data was as expected and that the rate limit dialog appeared. Also was able to verify the DB pruning manually in this fashion and saw old records drop away from the DB.
